contracts/access/manager/AccessManager.sol

```diff
@@ -0,0 +1,317 @@
+ * is connected to this contract as its manager, and 2) said function is decorated with the `restricted` modifier.
+ *
+ * There is a special team defined by default named "all" of which all accounts are considered members, and two special
+ * contract groups: 1) the "open" team, where all functions are allowed to the "all" team, and 2) the "closed" team,
+ * where no function is allowed to any team.
+ *
+ * Permissioning schemes and team and contract group assignments can be configured by the default admin. The contract
+ * includes {AccessControlDefaultAdminRules} by default to enforce security rules on this account. Additionally, it is
+ * expected that the account will be highly secured (e.g., a multisig or a well-configured DAO) as all the permissions
+ * of the managed system can be modified by it.
+ */
+contract AccessManager is IAccessManager, AccessControlDefaultAdminRules {
+    bytes32 _createdTeams;
+    mapping(address => bytes32) private _userTeams;
+    mapping(bytes32 => mapping(bytes4 => bytes32)) private _allowedTeams;
+    mapping(address => bytes32) private _contractGroup;
+
+    uint8 private constant _TEAM_ALL = 255;
+    bytes32 private constant _GROUP_UNSET = 0;
+    bytes32 private constant _GROUP_OPEN = "group:open";
+    bytes32 private constant _GROUP_CLOSED = "group:closed";
+
+    /**
+     * @dev Initializes an AccessManager with initial default admin and transfer delay.
+     */
+    constructor(
+        uint48 initialDefaultAdminDelay,
+        address initialDefaultAdmin
+    ) AccessControlDefaultAdminRules(initialDefaultAdminDelay, initialDefaultAdmin) {
+        createTeam(_TEAM_ALL, "all");
+    }
+
+    /**
+     * @dev Returns true if the caller can invoke on a target the function identified by a function selector.
+     * Entrypoint for {AccessManaged} contracts.
+     */
+    function canCall(address caller, address target, bytes4 selector) public view returns (bool) {
+        bytes32 group = getContractGroup(target);
+        bytes32 allowedTeams = getFunctionAllowedTeams(group, selector);
+        bytes32 callerTeams = getUserTeams(caller);
+        return callerTeams & allowedTeams != 0;
+    }
+
+    /**
+     * @dev Creates a new team with a team number that can be chosen arbitrarily but must be unused, and gives it a
+     * human-readable name. The caller must be the default admin.
+     *
+     * Emits {TeamUpdated}.
+     */
+    function createTeam(uint8 team, string memory name) public virtual onlyDefaultAdmin {
+        require(!_teamExists(team));
+        _setTeam(_createdTeams, team, true);
+        emit TeamUpdated(team, name);
+    }
+
+    /**
+     * @dev Updates an existing team's name. The caller must be the default admin.
+     */
+    function updateTeamName(uint8 team, string calldata name) public virtual onlyDefaultAdmin {
+        require(team != _TEAM_ALL && _teamExists(team));
+        emit TeamUpdated(team, name);
+    }
```

Comment:
What was the rational for needing to explicitly create teams?

Reply:
This gives the assurance that a team is "clean" and has no pre-existing members. I'm not entirely convinced that this is important. I was thinking of the situation where you're auditing a governance proposal or multisig proposal where there is a new team being defined that is granted permissions to a set of functions. You want to make sure that it is actually a new team that doesn't have pre-existing members and seeing the `createTeam` function call ensures that that's true (as long as the batch is atomic).

Do you agree that this is a benefit of having explicit creation? Do you see good reasons to remove explicit creation?

Something that could be argued is that contract groups have a similar problem with respect to being "clean", and those have no explicit creation, although they could have it too, but I think that would be more cumbersome operationally because many contracts will have ad hoc groups.